### PR TITLE
Truncate nodes list to maximum instance group size (1000) before adding to instance group

### PR DESCRIPTION
**Pull request by panslava**
This fixes a bug for clusters bigger than instance group max size, cause we were trying to add more than allowed number of nodes to instance group and gcp api returned an error

### CODE REVIEW

File: pkg/instances/instances_test.go [hunk 1]
```diff
@@ -116,6 +119,42 @@ func TestNodePoolSync(t *testing.T) {
 		t.Fatalf(
 			"Did not expect any calls, got %+v", fakeIGs.calls)
 	}
+
+	// KubeNodes: n1, n2, ..., n1000, n1001
+	// GCENodes:
```
Comment (kl52752) [inline pkg/instances/instances_test.go:124]: Can you add test where GCENodes are not empty and some might not be in KubeNodes list?

File: pkg/instances/instances_test.go [hunk 2]
```diff
@@ -116,6 +119,42 @@ func TestNodePoolSync(t *testing.T) {
...
+	// Try to add 1001 instances to the instance group. Should truncate to 1000 instances
+	fakeIGs = NewFakeInstanceGroups(map[string]IGsToInstances{
+		defaultZone: {
+			ig: sets.NewString(),
+		},
+	})
+	pool = newNodePool(fakeIGs, defaultZone)
+	pool.EnsureInstanceGroupsAndPorts(defaultNamer.InstanceGroup(), []int64{80})
+
+	fakeIGs.calls = []int{}
```
Comment (kl52752) [inline pkg/instances/instances_test.go:134]: can you explain why it is needed?

**Review by kl52752 — COMMENTED** [review]
tiny comments, beside that lgtm

File: pkg/instances/instances_test.go (hunk 1 above)
Comment (panslava) [reply pkg/instances/instances_test.go:124]: Done

File: pkg/instances/instances_test.go (hunk 2 above)
Comment (panslava) [reply pkg/instances/instances_test.go]: Not really needed for this test, removed

File: pkg/instances/instances_test.go [hunk 3]
```diff
@@ -41,6 +42,8 @@ func newNodePool(f *FakeInstanceGroups, zone string) NodePool {
 }
 
 func TestNodePoolSync(t *testing.T) {
```
Comment (cezarygerard) [inline pkg/instances/instances_test.go:44]: this should be a table test

**Review by cezarygerard — APPROVED** [review]
please see my 1 comment

File: pkg/instances/instances_test.go (hunk 3 above)
Comment (panslava) [reply pkg/instances/instances_test.go:44]: Done

**Comment by cezarygerard** [comment]
Slavik,

Thank you for refactoring the tests!

**Comment by cezarygerard** [comment]
@bowei ping  we need this PR for RBS launch

File: pkg/instances/instances.go [hunk 4]
```diff
@@ -334,6 +335,13 @@ func (i *Instances) Sync(nodes []string) (err error) {
 		}
 		kubeNodes := sets.NewString(nodes...)
 
+		// Individual InstanceGroup has a limit for 1000 instances in it.
+		// As a result, it's not possible to add more to it.
+		if len(kubeNodes) > flags.F.MaxIgSize {
```
Comment (bowei) [inline pkg/instances/instances.go:340]: Should this truncation be stable... e.g. should we sort the list before truncation?
I don't know if there is an explicit guarantee that the result of kubeNodes.List() is always going to return the same set of Nodes.

File: pkg/instances/instances.go [hunk 5]
```diff
@@ -334,6 +335,13 @@ func (i *Instances) Sync(nodes []string) (err error) {
...
+			klog.Warningf("Limiting number of VMs for InstanceGroup %s to %d", igName, flags.F.MaxIgSize)
```
Comment (bowei) [inline pkg/instances/instances.go:341]: I think we need to print more info here:

* Number of entries total 
* A sample of the VMs ignored (print at most the first 10)
* We should say "Nodes" instead of "VMs" (keep the naming consistent)
*

File: pkg/instances/instances.go (hunk 4 above)
Comment (panslava) [reply pkg/instances/instances.go:340]: That is a good question, I was curious about this as well

.List() actually "returns the contents as a sorted string slice."
https://github.com/kubernetes/ingress-gce/blob/master/vendor/k8s.io/apimachinery/pkg/util/sets/string.go#L170

Comment (panslava) [reply pkg/instances/instances.go:340]: To be sure, I've added to unit tests assertion

// call sync one more time and check that it will be no-op and will not cause any api calls

File: pkg/instances/instances.go (hunk 5 above)
Comment (panslava) [reply pkg/instances/instances.go]: Done

File: pkg/instances/instances.go (hunk 4 above)
Comment (bowei) [reply pkg/instances/instances.go:340]: Let's add a comment here so future reviewers don't have the same question

// List() will return a sorted list so the kubeNodesList truncation will have a stable set of nodes.

**Comment by bowei** [comment]
/approve

Ping back when you add the comment.

File: pkg/instances/instances.go (hunk 4 above)
Comment (panslava) [reply pkg/instances/instances.go:340]: Done

**Comment by panslava** [comment]
Okay, we discussed with Cezary and agreed on merging this, with hacky way of making the tests pass by doing 

`flags.F.MaxIgSize = 1000`

but will refactor this in other pull request

**Comment by panslava** [comment]
Request following this one, with refactoring https://github.com/kubernetes/ingress-gce/pull/1715

File: pkg/instances/instances.go [hunk 6]
```diff
@@ -334,6 +335,23 @@ func (i *Instances) Sync(nodes []string) (err error) {
...
 
+		// Individual InstanceGroup has a limit for 1000 instances in it.
+		// As a result, it's not possible to add more to it.
+		if len(kubeNodes) > flags.F.MaxIgSize {
+			// List() will return a sorted list so the kubeNodesList truncation will have a stable set of nodes.
+			kubeNodesList := kubeNodes.List()
+
+			// Store first 10 truncated nodes for logging
+			truncatedNodesSample := kubeNodesList[flags.F.MaxIgSize:]
+			maxTruncatedNodesSampleSize := 10
```
Comment (bowei) [inline pkg/instances/instances.go:346]: suggestion:

```
truncateForLog := func(l []string) []string {
  if len(l) <= 10 { return l }
  return l[:10]
}
```

klog.Warning(..., truncateForLog(kubeNodesList))

This way we don't pollute the variable state space and make it very clear what the intent is...
